agent/stats/engine.go

```diff
@@ -0,0 +1,446 @@
+		if !exists {
+			log.Debug("Could not map task to definition", "task", taskArn)
+			continue
+		}
+
+		taskMetrics = append(taskMetrics, &ecstcs.TaskMetric{
+			TaskArn:               &taskArn,
+			TaskDefinitionFamily:  &taskDef.family,
+			TaskDefinitionVersion: &taskDef.version,
+			ContainerMetrics:      containerMetrics,
+		})
+	}
+
+	if len(taskMetrics) == 0 {
+		// Not idle. Expect taskMetrics to be there.
+		return nil, nil, errors.New("No task metrics to report")
+	}
+
+	return engine.metricsMetadata, taskMetrics, nil
+}
+
+func (engine *DockerStatsEngine) isIdle() bool {
+	return len(engine.tasksToContainers) == 0
+}
+
+// initDockerClient initializes engine's docker client.
+func (engine *DockerStatsEngine) initDockerClient() error {
+	if engine.client == nil {
+		client, err := ecsengine.NewDockerGoClient()
+		if err != nil {
+			return err
+		}
+		engine.client = client
+	}
+
+	return nil
+}
+
+// addExistingContainers lists existing containers and adds them to the engine.
+func (engine *DockerStatsEngine) addExistingContainers() error {
+	listContainersResponse := engine.client.ListContainers(false)
+	if listContainersResponse.Error != nil {
+		return listContainersResponse.Error
+	}
+
+	for _, containerID := range listContainersResponse.DockerIds {
+		engine.AddContainer(containerID)
+	}
+
+	return nil
+}
+
+// openEventStream initializes the channel to receive events from docker client's
+// event stream.
+func (engine *DockerStatsEngine) openEventStream() error {
+	events, err := engine.client.ContainerEvents(engine.ctx)
+	if err != nil {
+		return err
+	}
+	engine.events = events
+	return nil
+}
+
+// handleDockerEvents must be called after openEventstream; it processes each
+// event that it reads from the docker event stream.
+func (engine *DockerStatsEngine) handleDockerEvents() {
+	for {
+		select {
+		case <-engine.ctx.Done():
+			return
+		case event := <-engine.events:
+			log.Debug("Handling an event: ", "container", event.DockerId, "status", event.Status.String())
+			switch event.Status {
+			case api.ContainerRunning:
+				engine.AddContainer(event.DockerId)
+			case api.ContainerStopped:
+				engine.RemoveContainer(event.DockerId)
+			default:
+				log.Debug("Ignoring event for container", "id", event.DockerId, "status", event.Status)
+			}
+		}
+	}
+	log.Crit("Docker event stream closed unexpectedly")
+}
+
+// newDockerContainerMetadataResolver returns a new instance of DockerContainerMetadataResolver.
+func newDockerContainerMetadataResolver(taskEngine ecsengine.TaskEngine) (*DockerContainerMetadataResolver, error) {
+	dockerTaskEngine, ok := taskEngine.(*ecsengine.DockerTaskEngine)
+	if !ok {
+		// Error type casting docker task engine.
+		return nil, errors.New("Could not load docker task engine")
+	}
+
+	resolver := &DockerContainerMetadataResolver{
+		dockerTaskEngine: dockerTaskEngine,
+	}
+
+	return resolver, nil
+}
+
+// setMetricCollectionFlag reads the ECS_DISABLE_METRICS env variable and
```

Comment:
I think I would prefer that `ECS_DISABLE_METRICS` were part of the `config` package we use elsewhere.  Capturing it there lets us handle it consistently with the rest of the configuration.